editors/code/src/inlay_hints.ts

```diff
@@ -1,162 +1,319 @@
+import * as lc from "vscode-languageclient";
 import * as vscode from 'vscode';
 import * as ra from './rust-analyzer-api';
 
 import { Ctx } from './ctx';
-import { log, sendRequestWithRetry } from './util';
+import { sendRequestWithRetry, assert } from './util';
 
 export function activateInlayHints(ctx: Ctx) {
-    const hintsUpdater = new HintsUpdater(ctx);
+    const hintsUpdater = new HintsUpdater(ctx.client);
+
     vscode.window.onDidChangeVisibleTextEditors(
-        async _ => hintsUpdater.refresh(),
+        visibleEditors => hintsUpdater.refreshVisibleRustEditors(
+            visibleEditors.filter(isRustTextEditor)
+        ),
         null,
         ctx.subscriptions
     );
 
     vscode.workspace.onDidChangeTextDocument(
-        async event => {
-            if (event.contentChanges.length === 0) return;
-            if (event.document.languageId !== 'rust') return;
-            await hintsUpdater.refresh();
+        ({ contentChanges, document }) => {
+            if (contentChanges.length === 0) return;
+            if (!isRustTextDocument(document)) return;
+
+            hintsUpdater.refreshRustDocument(document);
```

Comment:
We should refresh all visible editors: change to one file might affect hints in another file.

Reply:
Done! https://github.com/rust-analyzer/rust-analyzer/pull/3378/commits/995ad75e23e788c1ed63b515d5e28045cf55438d